### ISSUE TITLE
Functional Preprocessing and Augmentation API

### ISSUE THREAD

**Issue by ahundt** [fchollet/keras#6655]
Edit: A new example proposal + details is in https://github.com/fchollet/keras/pull/6891#issuecomment-307659460.

Discussion about a dense prediction API such as image segmentation in https://github.com/fchollet/keras/issues/6538 brought up the possibility of a functional preprocessing API, which could move preprocessing steps into the Keras backend APIs and generalize preprocessing to more network designs. [Dropout](https://github.com/fchollet/keras/blob/master/keras/layers/core.py#L72) provides a precedent for such augmentation layers.

If the layers could be designed much like dropout, and I would expect them to be configured so that the augmentation operations could be applied identically to one or more image inputs as well as one or more image label data, useful for dense prediction tasks.

This could have the advantages of being easy to use, easily applied consistently for arbitrary data inputs, and make it possible to use the [TF backend image augmentation APIs](https://www.tensorflow.org/api_guides/python/image) thus improving performance.

What would be the pros/cons and barriers to a Functional Preprocessing API?

Example usage, label augmentation optional for dense prediction tasks:

```python
L = InputLabel(...)
input = Input(...)

# augmentation
x,L = Flip(axis=0)([input,L])
x,L = Flip(axis=1)([x,L])
x,L = Zoom(range=[0.5,2])([x,L])
x,L = Transform(matrix=[1,0,0,0,1,0,0,0,1], interpolation='bilinear')([x,L])
x,L = RandomCrop(fill_mode='constant', cval=0.)([x,L])

# network definition
x = Conv2D(inter_channel, (1, 1), kernel_initializer='he_uniform', padding='same', use_bias=False,
                   kernel_regularizer=l2(weight_decay))(x)

# ...snip...
model = Model(inputs=input, outputs=x, labels=L)
model.compile(Adam(), loss='categorical_crossentropy',metrics=['acc'])
# ...snip...
model.fit()
```

**Comment by Dref360** [comment]
Is there anything like the tf preprocessing for theano? Otherwise, I guess it's a deal-breaker for theano users.

**Comment by ahundt** [comment]
@Dref360 I don't know, perhaps plain python APIs could be run on the theano side? They already have the sequential api you mentioned so adaptation may not be too hard, where would a feature request for theano be submitted?

**Comment by ahundt** [comment]
Edit: A new example proposal + details is in https://github.com/fchollet/keras/pull/6891#issuecomment-307659460.

**Comment by ahundt** [comment]
If you're using the TensorFlow backend, you can just use input tensors by following the [mnist_tfrecord.py](https://github.com/fchollet/keras/blob/master/examples/mnist_tfrecord.py) example, and before passing your tensors to the Input layer apply TensorFlow preprocessing ops such as [tf.random_crop](https://www.tensorflow.org/api_docs/python/tf/random_crop). 

Since there is a way to accomplish this, I'm closing the issue.
